{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One operation I often do is fuse protein together... This is the automated version."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prevent pymol from launching in normal mode.\n",
    "if __name__ == '__main__':\n",
    "    pymol_argv = ['pymol', '-qc']\n",
    "else:\n",
    "    import __main__\n",
    "    __main__.pymol_argv = ['pymol', '-qc']\n",
    "\n",
    "import pymol\n",
    "pymol.finish_launching()\n",
    "\n",
    "from pprint import PrettyPrinter\n",
    "pprint = PrettyPrinter().pprint\n",
    "\n",
    "\n",
    "import os, requests, time, json, uuid\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from IPython.display import display, HTML, Image\n",
    "from Bio.SeqUtils import seq3\n",
    "from collections import namedtuple\n",
    "from warnings import warn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import protein #venus protein...\n",
    "Protein = protein.ProteinCore\n",
    "protein = 'P27708'\n",
    "p = Protein(uniprot=protein).gload()\n",
    "print([s.id for s in p.pdbs])\n",
    "print([s.id for s in p.swissmodel])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdbs = ['1C0F_S', '1C0G_S', '1D4X_G', '1DEJ_S', '1EQY_S', '1ESV_S', '1H1V_G', '1KCQ_A', '1MDU_A', '1NLV_G', '1NM1_G', '1NMD_G', '1P8X_A', '1P8Z_G', '1SOL_A', '1T44_G', '1YAG_G', '1YVN_G', '2FF3_A', '2FF6_G', '2FH1_A', '2FH2_A', '2FH3_A', '2FH4_A', '3A5L_S', '3A5M_S', '3A5N_S', '3A5O_S', '3CI5_G', '3CIP_G', '3CJB_G', '3CJC_G', '3FFK_A', '3FFN_A', '3TU5_B', '4PKG_G', '4PKH_B', '4PKI_G', '4S10_C', '4Z94_G', '5DD2_A', '5FAE_A', '5FAF_A', '5H3M_A', '5H3N_A', '5O2Z_A', '5UBO_S', '1C0F_S', '1C0G_S', '1D4X_G', '1DEJ_S', '1EQY_S', '1ESV_S', '1H1V_G', '1KCQ_A', '1MDU_A', '1NLV_G', '1NM1_G', '1NMD_G', '1P8X_A', '1P8Z_G', '1SOL_A', '1T44_G', '1YAG_G', '1YVN_G', '2FF3_A', '2FF6_G', '2FH1_A', '2FH2_A', '2FH3_A', '2FH4_A', '3A5L_S', '3A5M_S', '3A5N_S', '3A5O_S', '3CI5_G', '3CIP_G', '3CJB_G', '3CJC_G', '3FFK_A', '3FFN_A', '3TU5_B', '4PKG_G', '4PKH_B', '4PKI_G', '4S10_C', '4Z94_G', '5DD2_A', '5FAE_A', '5FAF_A', '5H3M_A', '5H3N_A', '5O2Z_A', '5UBO_S']\n",
    "swissmodel = ['5c5b926d732c1ed759c144a4', '5c5b926d732c1ed759c144a8', '5c5b926d732c1ed759c144ac', '5c5b926d732c1ed759c144b0', '5bf80f5102efd0fe0bb1b4b4', '5bf7e47602efd06ec1fcae16', '5be49a9b02efd0bc6679f738', '5bf80ed402efd0fc662447d4', '5be49a2302efd0bb1a29e5b0', '5be4ab9402efd0e942408106', '5bf7e33a02efd06a62865646', '5bf7de1502efd05a351620ea', '5be4a56f02efd0d9fdee17f9', '5be460af02efd01e3e55ae46', '5be4abc602efd0ea9b334ba3', '5bf7e07b02efd061f471a1f2', '5bf8103b02efd001eeb29f09', '5bf810ac02efd0035d2f8b94', '5be4749102efd057e999f715', '5be4af9a02efd0f45721bcf9', '5bf812c702efd00a2ba4e36b', '5be4981602efd0b60bc038fa', '5bf7e0da02efd06349374fc1', '5bf80c9802efd0f578c9c9d7', '5be476ba02efd05d6b67854b', '5bf7e77d02efd0771630aeea', '5be49d4c02efd0c2e5c15fd1', '5be4631402efd025857aacdf', '5bf7df0e02efd05d991ec14b', '5bf8090202efd0ebaa275be0', '5c480a4c02efd06f714f182d', '5bf7f81002efd0ab54f3b027', '5be47e3902efd0719430de91', '5bf8100202efd001eeb29a7d', '5bf7ffee02efd0c71c3987ad', '5bf7e9ea02efd07e64245202', '5bf7e67902efd074e05524f4', '5bf7e77202efd0771630ae25', '5bf7eccf02efd08642607de0', '5bf7ed3b02efd088b9a29267', '5bf7fa2602efd0b19ad8c0c0', '5bf7fc7502efd0bbef60ea6e', '5bf8038202efd0d87451aba6', '5be47d1302efd06e7ae26add', '5be4951702efd0adb5f15cb2', '5bf7edfb02efd08b2ba2b519', '5bf8018702efd0ceeac99171', '5bf8020c02efd0d019ae92ce', '5be4ae2202efd0f177fb6a1d', '5bf805fb02efd0e07b4842c5', '5bf80acb02efd0f0bcf33a0d', '5bf80e5d02efd0fb527e7bb6']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# switch to dock9 as it has non overlapping regions.\n",
    "uniprot = 'Q9BZ29' #'DOCK9'\n",
    "pdbs = [{'id': '1WG7_A', 'x': 165, 'y': 301}, {'id': '2WM9_A', 'x': 1605, 'y': 2067}, {'id': '2WMN_A', 'x': 1605, 'y': 2067}, {'id': '2WMO_A', 'x': 1605, 'y': 2067}, {'id': '1WG7_A', 'x': 165, 'y': 301}, {'id': '2WM9_A', 'x': 1605, 'y': 2067}, {'id': '2WMN_A', 'x': 1605, 'y': 2067}, {'id': '2WMO_A', 'x': 1605, 'y': 2067}]\n",
    "swissmodel = ['5c60bed13b85d1779bf56e74', '5c60bed13b85d1779bf56e7c', '5c60bed13b85d1779bf56e78', '5c60bed13b85d1779bf56e80', '5be4861702efd085a65cee63', '5bf7f3f502efd09e53d2d042', '5bf80fb702efd0ffaa6184cf', '5bf8111202efd004641735f8']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pdbs = ['1C0F_S', '1C0G_S', '1D4X_G', '1DEJ_S', '1EQY_S', '1ESV_S', '1H1V_G', '1KCQ_A', '1MDU_A', '1NLV_G', '1NM1_G', '1NMD_G', '1P8X_A', '1P8Z_G', '1SOL_A', '1T44_G', '1YAG_G', '1YVN_G', '2FF3_A', '2FF6_G', '2FH1_A', '2FH2_A', '2FH3_A', '2FH4_A', '3A5L_S', '3A5M_S', '3A5N_S', '3A5O_S', '3CI5_G', '3CIP_G', '3CJB_G', '3CJC_G', '3FFK_A', '3FFN_A', '3TU5_B', '4PKG_G', '4PKH_B', '4PKI_G', '4S10_C', '4Z94_G', '5DD2_A', '5FAE_A', '5FAF_A', '5H3M_A', '5H3N_A', '5O2Z_A', '5UBO_S', '1C0F_S', '1C0G_S', '1D4X_G', '1DEJ_S', '1EQY_S', '1ESV_S', '1H1V_G', '1KCQ_A', '1MDU_A', '1NLV_G', '1NM1_G', '1NMD_G', '1P8X_A', '1P8Z_G', '1SOL_A', '1T44_G', '1YAG_G', '1YVN_G', '2FF3_A', '2FF6_G', '2FH1_A', '2FH2_A', '2FH3_A', '2FH4_A', '3A5L_S', '3A5M_S', '3A5N_S', '3A5O_S', '3CI5_G', '3CIP_G', '3CJB_G', '3CJC_G', '3FFK_A', '3FFN_A', '3TU5_B', '4PKG_G', '4PKH_B', '4PKI_G', '4S10_C', '4Z94_G', '5DD2_A', '5FAE_A', '5FAF_A', '5H3M_A', '5H3N_A', '5O2Z_A', '5UBO_S']\n",
    "swissmodel = ['5c5b926d732c1ed759c144a4', '5c5b926d732c1ed759c144a8', '5c5b926d732c1ed759c144ac', '5c5b926d732c1ed759c144b0', '5bf80f5102efd0fe0bb1b4b4', '5bf7e47602efd06ec1fcae16', '5be49a9b02efd0bc6679f738', '5bf80ed402efd0fc662447d4', '5be49a2302efd0bb1a29e5b0', '5be4ab9402efd0e942408106', '5bf7e33a02efd06a62865646', '5bf7de1502efd05a351620ea', '5be4a56f02efd0d9fdee17f9', '5be460af02efd01e3e55ae46', '5be4abc602efd0ea9b334ba3', '5bf7e07b02efd061f471a1f2', '5bf8103b02efd001eeb29f09', '5bf810ac02efd0035d2f8b94', '5be4749102efd057e999f715', '5be4af9a02efd0f45721bcf9', '5bf812c702efd00a2ba4e36b', '5be4981602efd0b60bc038fa', '5bf7e0da02efd06349374fc1', '5bf80c9802efd0f578c9c9d7', '5be476ba02efd05d6b67854b', '5bf7e77d02efd0771630aeea', '5be49d4c02efd0c2e5c15fd1', '5be4631402efd025857aacdf', '5bf7df0e02efd05d991ec14b', '5bf8090202efd0ebaa275be0', '5c480a4c02efd06f714f182d', '5bf7f81002efd0ab54f3b027', '5be47e3902efd0719430de91', '5bf8100202efd001eeb29a7d', '5bf7ffee02efd0c71c3987ad', '5bf7e9ea02efd07e64245202', '5bf7e67902efd074e05524f4', '5bf7e77202efd0771630ae25', '5bf7eccf02efd08642607de0', '5bf7ed3b02efd088b9a29267', '5bf7fa2602efd0b19ad8c0c0', '5bf7fc7502efd0bbef60ea6e', '5bf8038202efd0d87451aba6', '5be47d1302efd06e7ae26add', '5be4951702efd0adb5f15cb2', '5bf7edfb02efd08b2ba2b519', '5bf8018702efd0ceeac99171', '5bf8020c02efd0d019ae92ce', '5be4ae2202efd0f177fb6a1d', '5bf805fb02efd0e07b4842c5', '5bf80acb02efd0f0bcf33a0d', '5bf80e5d02efd0fb527e7bb6']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 247,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "##resn = seq3(self.seq[self.end]).upper()\n",
    "\n",
    "_Coords = namedtuple('Coords',['name', 'x','y','z'])\n",
    "\n",
    "class Coords(_Coords):\n",
    "    def to_np(self, coords):\n",
    "        return np.array([coords.x, coords.y, coords.z])\n",
    "\n",
    "def coords_to_np(self, coords):\n",
    "        np.array([coords.x, coords.y, coords.z])\n",
    "\n",
    "####################################################################################\n",
    "class Workshop:\n",
    "    \"\"\"\n",
    "    >>> ws = Workshop(pdbs, debug).order().save()\n",
    "    debug: verbosity boolean\n",
    "    pdbs: [{'x': stated_start_in_uniprot,\n",
    "            'y': stated_end_in_uniprot,\n",
    "            'id': 'code_chain', \n",
    "            'tier': opt. int. low => max importance (e.g. crystal), high => lower importance (e.g. models)\n",
    "            'description': opt. 'Not used.'}]\n",
    "    In the case of PDBs the metadata is taken in order to find where the true start is without cloning scars.\n",
    "    The code can be a PDB code. or 'PDB:code'  or 'SWISSMODEL:id' or 'LOCAL:filename' (without underscores!)\n",
    "    For PDB codes, the metadata is fetch to verify what the true values are.\n",
    "    See model for more about the number madness\n",
    "    order() readies and aligns or projects the non-redudant models.\n",
    "    the init step actually calls `.categorise()` which categorises the models based on redundancy.\n",
    "    ready_models() loads models\n",
    "    align_models() aligns overlapping ones\n",
    "    project_models() places models with gaps along the x-axis.\n",
    "    `ws.models` list of models.\n",
    "    Note that there is not a create step to fuse the parts as simply saving as pdb will do it.\n",
    "    But saving as pse to check things is nice.\n",
    "    `pymol.exporting.multisave('tmp.pdb')` is a handy command if there is a version mismathc with pymol.\n",
    "    \n",
    "    Model has many bound methods that use pymol cmds, that affect only one model, say `protein.roll(40)`,\n",
    "    while workshop uses more thatn one say, `ws.roll_free(protein_A, protein_B)`.\n",
    "    Some methods seems repeated but aren't. the `_models` operate on all.\n",
    "    ws.ready_models() --> model.fetch_n_clean()\n",
    "    ws.align_models() --> ws.align(A,B)\n",
    "    ws.project_models() --> ws.project(A,B)\n",
    "    \"\"\"\n",
    "    def __init__(self, pdbs, debug=False):\n",
    "        self.debug = bool(debug)\n",
    "        self.pdbs = [{'tier': 1, 'description': 'NA', **p} for p in pdbs]\n",
    "        self.metadata = {}\n",
    "        self.fetch_metadata()\n",
    "        if self.debug:\n",
    "            print(self.metadata)\n",
    "        self.models = []\n",
    "        self.catergorise()\n",
    "        ### pymol operations.\n",
    "        self.last_het_index = 9000\n",
    "        self.joints = [] ##{'joint': 'resi_fore': int, 'fore_model': 'aft'..., 'type': 'end|alignment|projection', 'overlap' distance 'note'}\n",
    "        \n",
    "    def log_joint(self, form='ERROR', fore=None, aft=None):\n",
    "        if fore is None and aft is None:\n",
    "            raise ValueError\n",
    "        elif fore is None: #nothing before\n",
    "            self.joints.append({'type': 'end',\n",
    "                               'aft_resi_start': aft.start,\n",
    "                               'aft_source': f'{aft.mode}:{aft.code}'})\n",
    "        elif aft is None: #nothing after\n",
    "            self.joints.append({'type': 'end',\n",
    "                               'fore_resi_start': fore.end,\n",
    "                               'fore_source': f'{fore.mode}:{fore.code}'})\n",
    "        else:\n",
    "            f = fore.get_resi_coords(fore.end)['C']\n",
    "            a = aft.get_resi_coords(aft.start)['N']\n",
    "            d = fore.distance(f, a)\n",
    "            o = pymol.cmd.overlap(f\"model {fore.name}\", f\"model {aft.name}\")\n",
    "            self.joints.append({'type': form,\n",
    "                               'fore_resi_end': fore.end,\n",
    "                               'fore_source': f'{fore.mode}:{fore.code}',\n",
    "                               'aft_resi_start': aft.start,\n",
    "                               'aft_source': f'{aft.mode}:{aft.code}',\n",
    "                               'distance': d,\n",
    "                               'overlap': o\n",
    "                               })\n",
    "        return self\n",
    "    \n",
    "    def order(self):\n",
    "        \"\"\"\n",
    "        clear pymol, ready models, align overlapping, project gaps.\n",
    "        \"\"\"\n",
    "        pymol.cmd.delete('all')\n",
    "        self.ready_models()\n",
    "        pymol.cmd.delete('cylinder*')\n",
    "        if self.debug:\n",
    "            self.show_load_checks()\n",
    "            self.show_before()  \n",
    "        if self.debug:\n",
    "            pymol.cmd.zoom(self.models[0].name)\n",
    "            self.show_pose('mid.png')\n",
    "            print('### Aligning!')\n",
    "        self.align_models()\n",
    "        if self.debug:\n",
    "            print('### Projecting!')\n",
    "        self.project_models()\n",
    "        self.log_joint('end', aft=self.models[0])\n",
    "        self.log_joint('end', fore=self.model[-1])\n",
    "        if self.debug:\n",
    "            self.show_after()\n",
    "        return self\n",
    "            \n",
    "    def save(self, filename='fusion.pdb'):\n",
    "        pymol.cmd.save(filename)\n",
    "        return self\n",
    "            \n",
    "    def align_models(self):\n",
    "        \"\"\"\n",
    "        Fuses all overlapping models. The name will be a uuid of 6 letters\n",
    "        \"\"\"\n",
    "        for i in range(1,len(self.models)):\n",
    "            fore = self.models[i - 1]\n",
    "            aft = self.models[i]\n",
    "            if fore.end > aft.start:\n",
    "                self.align(fore, aft)\n",
    "                self.log_joint('alignment', fore, aft)\n",
    "                #changing aft to a fusion.\n",
    "                self.models[i].start = fore.start\n",
    "                self.models[i].length = fore.length + aft.length\n",
    "                new_name = uuid.uuid1().hex.replace('-','')[:6]\n",
    "                pymol.cmd.create(new_name, f'{fore.name} or {aft.name}')\n",
    "                pymol.cmd.delete(fore.name)\n",
    "                pymol.cmd.delete(aft.name)\n",
    "                self.models[i].name = new_name\n",
    "                self.models[i-1] = None ## kill fore from list.\n",
    "        self.models = [m for m in self.models if m is not None]\n",
    "        return self\n",
    "        \n",
    "    def project_models(self):\n",
    "        \"\"\"\n",
    "        places models with gaps along the x-axis.\n",
    "        \"\"\"\n",
    "        for m in self.models: ##no idea why this is needed.\n",
    "            m.angle_fix()\n",
    "        for i in range(1,len(self.models)):\n",
    "            fore = self.models[i - 1]\n",
    "            aft = self.models[i]    \n",
    "            self.project(fore, aft)\n",
    "            self.roll_free(fore, aft)\n",
    "            self.log_joint('projection', fore, aft)\n",
    "        return self\n",
    "        \n",
    "    def show_pose(self, filename='temp.png'):\n",
    "        \"\"\"\n",
    "        Jupyter notebook specific.\n",
    "        \"\"\"\n",
    "        if '.png' not in filename:\n",
    "            filename += '.png'\n",
    "        if os.path.exists(filename):\n",
    "            os.remove(filename)\n",
    "        pymol.cmd.png(filename)\n",
    "        while not os.path.exists(filename):\n",
    "            time.sleep(1)\n",
    "        display(Image(filename=filename))\n",
    "    \n",
    "    def show_before(self):\n",
    "        self.add_xyz()\n",
    "        modsele = f'{self.models[0].sele(self.models[0].end)} or {self.models[1].sele(self.models[1].start)} or {self.models[1].sele(self.models[1].end)}'\n",
    "        pymol.cmd.color('yellow',modsele)\n",
    "        pymol.cmd.show(representation=\"sticks\", selection=modsele)\n",
    "        pymol.cmd.zoom(modsele)\n",
    "        print('Before')\n",
    "        self.show_pose('before.png')\n",
    "        \n",
    "    def show_after(self):\n",
    "        print('after')\n",
    "        pymol.cmd.zoom('all')\n",
    "        self.show_pose('after.png')\n",
    "        pymol.cmd.save('after.pdb', 'all')\n",
    "    \n",
    "\n",
    "    def show_load_checks(self):\n",
    "        pymol.cmd.zoom('all')\n",
    "        for model in self.models:\n",
    "            pymol.cmd.color('cyan',model.name)\n",
    "            pymol.cmd.color('red',f'{model.name} and chain {model.chain}')\n",
    "            modsele = model.sele(f'{model.start}-{model.end}')\n",
    "            pymol.cmd.color('white',modsele)\n",
    "        self.show_pose('all.png')\n",
    "        \n",
    "    def roll_free(self, fore, aft, cycle=1):\n",
    "        if cycle == 0:\n",
    "            return self\n",
    "        if self.has_overlap(fore, aft):\n",
    "            for i in range(6):\n",
    "                print(f'Overlap between {fore.name} and {aft.name}')\n",
    "                aft.roll(60)\n",
    "                if not self.has_overlap(fore, aft):\n",
    "                    return self\n",
    "            else:\n",
    "                print(f'Failed cycle {cycle}')\n",
    "                self.project(fore, aft, extra_d=5 * cycle)\n",
    "                self.roll_free(fore, aft, cycle = cycle + 1)\n",
    "        return self\n",
    "                \n",
    "    def align(self, fore, aft, cycle=0):\n",
    "        mobile = aft.sele(f\"{aft.start}-{fore.end}\")\n",
    "        m_match = self.safe_select(mobile)\n",
    "        target = fore.sele(f\"{aft.start}-{fore.end}\")\n",
    "        t_match = self.safe_select(target)\n",
    "        if m_match == 0 or t_match == 0:\n",
    "            warn(f'Possible major issue...\\n{mobile} ... {m_match}\\n{target} ... {t_match}')\n",
    "            return self\n",
    "        pymol.cmd.align(mobile, target)\n",
    "        scores = {}\n",
    "        if fore.tier <=aft.tier:\n",
    "            ranger = range(fore.end, aft.start-1, -1)\n",
    "        else:\n",
    "            ranger = range(aft.start, fore.end+1, +1)\n",
    "        for i in ranger:\n",
    "            pymol.cmd.pair_fit(aft.sele(i, 'N'), fore.sele(i, 'N'),\n",
    "                               aft.sele(i, 'CA'), fore.sele(i, 'CA'),\n",
    "                               aft.sele(i, 'C'), fore.sele(i, 'C'))\n",
    "            if self.debug:\n",
    "                print(f'... {mobile} --> {target}')\n",
    "                pymol.cmd.zoom(mobile)\n",
    "                pymol.cmd.color('red',mobile)\n",
    "                pymol.cmd.color('green',target)\n",
    "                self.show_pose('align.png')\n",
    "            if cycle > 0:\n",
    "                aft.roll(60 * cycle)\n",
    "            o = pymol.cmd.overlap(fore.sele(f\"{fore.start}-{i}\"), aft.sele(f\"{i+1}-{aft.end}\"))\n",
    "            if o == 0:\n",
    "                break\n",
    "            else:\n",
    "                if self.debug:\n",
    "                    print(f'!!! There is an overlap of {o}!')\n",
    "                scores[i] = o\n",
    "        else:\n",
    "            i = list(scores.keys())[list(scores.values()).index(min(scores.values()))]\n",
    "            if scores[i] < 100:\n",
    "                if self.debug:\n",
    "                    print(f'Aligning via residue {i}. score={scores[i]}')\n",
    "                pymol.cmd.pair_fit(fore.sele(i, 'N'), aft.sele(i, 'N'),\n",
    "                                       fore.sele(i, 'CA'), aft.sele(i, 'CA'),\n",
    "                                       fore.sele(i, 'C'), aft.sele(i, 'C'))\n",
    "                if cycle > 0:\n",
    "                    aft.roll(60 * cycle)\n",
    "            else:\n",
    "                print('Good alignment impossible.')\n",
    "                if cycle < 6:\n",
    "                    align(self, fore, aft, cycle=cycle + 1)\n",
    "                else:\n",
    "                    raise Exception('Really?? Overlap after alignment even with rotations?? Sigh.')\n",
    "        if fore.tier <=aft.tier:\n",
    "            pymol.cmd.remove(fore.sele(f'{i+1}-{fore.end}'))\n",
    "            fore.end = i\n",
    "            pymol.cmd.remove(aft.sele(f'{aft.start}-{i}'))\n",
    "            aft.start = i + 1\n",
    "        else:\n",
    "            pymol.cmd.remove(fore.sele(f'{i}-{fore.end}'))\n",
    "            fore.end = i - 1\n",
    "            pymol.cmd.remove(aft.sele(f'{aft.start}-{i-1}'))\n",
    "            aft.start = i\n",
    "        return self\n",
    "        \n",
    "    def fetch_metadata(self):\n",
    "        if self.debug: print('## Fetching metadata')\n",
    "        ### fetch the metadata\n",
    "        for entry in self.pdbs:\n",
    "            mode, code, chain = Model.split_id(entry)\n",
    "            if mode == 'PDB':\n",
    "                reply = requests.get(f'http://www.ebi.ac.uk/pdbe/api/pdb/entry/molecules/{code}').json()\n",
    "                self.metadata[code] = reply[code.lower()]\n",
    "            else:\n",
    "                self.metadata[code] = {}\n",
    "        return self\n",
    "    \n",
    "    def catergorise(self):\n",
    "        \"\"\"\n",
    "        Clusters the proteins\n",
    "        \"\"\"\n",
    "        if self.debug: print('## Running categorisation')\n",
    "        ## find longest!\n",
    "        tdata = []\n",
    "        from collections import defaultdict\n",
    "        starters = defaultdict(list)\n",
    "        enders = defaultdict(list)\n",
    "        for entry in self.pdbs:\n",
    "            print(entry)\n",
    "            mode, code, chain = Model.split_id(entry)\n",
    "            if mode == 'PDB':\n",
    "                c_entity = [entity for entity in self.metadata[code] if chain in entity['in_chains']][0]\n",
    "                mappings = c_entity['source'][0]['mappings']\n",
    "                if len(c_entity['source'][0]['mappings']) > 1:\n",
    "                    raise ValueError('MULTIPLE MAPPINGS!')\n",
    "                s = mappings[0]['start']['residue_number']\n",
    "                e = mappings[0]['end']['residue_number']\n",
    "                seq = c_entity['sequence']\n",
    "            else:\n",
    "                #Model.infer_metadata(entry)\n",
    "                #burden on user.\n",
    "                s = 1\n",
    "                e = entry['y'] - entry['x'] + 1\n",
    "                seq = None\n",
    "            tier = float(entry['tier'])\n",
    "            tdata.append({'code': code,\n",
    "                          'mode': mode,\n",
    "                          'chain': chain,\n",
    "                          'tier': tier,\n",
    "                          'x': entry['x'],\n",
    "                          'y': entry['y'],\n",
    "                          'seq': seq,\n",
    "                         'start': s,\n",
    "                         'end': e,\n",
    "                         'length': e - s})\n",
    "        df = pd.DataFrame.from_records(tdata).sort_values(['x', 'length', 'tier'], ascending=[True, False, True])\n",
    "        if self.debug:\n",
    "            print('There are all the chains:')\n",
    "            display(df)\n",
    "        ## thin!\n",
    "        new = df.iloc[0]\n",
    "        get_similars = lambda: list(set(df.loc[(df.y <=  new.y) & (df.x >=  new.x)].code.values) - {new.code})\n",
    "        new.at['similars']= get_similars()\n",
    "        o = df.loc[df.y > new.y]\n",
    "        parts = [new]\n",
    "        old = new\n",
    "        while len(o):\n",
    "            new = o.iloc[0]\n",
    "            new.at['similars'] = get_similars()\n",
    "            if new.x < old.y: # mark overlap for debug.\n",
    "                new.at['subtender'] = old.code\n",
    "            ## curious case of better models within makes this weird.\n",
    "            # so something is better is pulled out of the similars (i.e. smaller so excluded.) group.\n",
    "            # but to be better it has to be of a lower tier and not present in the selection.\n",
    "            betters = df[(df.code.isin(new.similars)) & (df.tier < new.tier) & (~df.code.isin([p.code for p in parts]))]\n",
    "            bi = len(betters.index)\n",
    "            if bi == 0:\n",
    "                parts.append(new)  ## has to happen!!\n",
    "            else:\n",
    "                better = betters.iloc[0]\n",
    "                print(f'Nested case {new.code} and {betters.code.values}')\n",
    "                ## dangerous hack. It assumes that a good match can be found and won't overrun.\n",
    "                ## as the priority differs it will chew off one side then the other, hence the double.\n",
    "                if better.x > new.x:\n",
    "                    parts.append(new)\n",
    "                parts.append(better)\n",
    "                if better.y < new.y:\n",
    "                    parts.append(new)\n",
    "            ## cleanup.\n",
    "            o = o.loc[o.y > new.y]\n",
    "            old = new\n",
    "        if self.debug:\n",
    "            print('These are the non-redudant chains:')\n",
    "            display(pd.DataFrame(parts))\n",
    "        self.models = [Model(workshop=self, metadata= self.metadata[row.code], **row) for row in parts]\n",
    "        return self\n",
    "    \n",
    "    def ready_models(self):\n",
    "        \"\"\"\n",
    "        Runs model.fetch_n_clean() on all models.\n",
    "        \"\"\"\n",
    "        if self.debug: print('## Reading models')\n",
    "        for model in self.models:\n",
    "            model.fetch_n_clean()\n",
    "        self.models[0].camera_fix() \n",
    "        return self\n",
    "    \n",
    "    def project(self, fore, aft, extra_d = 0):\n",
    "        \"\"\"extra_d is to move it away when there is on clean rotational solution.\"\"\"\n",
    "        aft.angle_fix()\n",
    "        (ref, pointer, θ, φ, internal_d) = fore.angles\n",
    "        (ref_aft, pointer_aft, θ_aft, φ_aft, internal_d_aft) = aft.angles\n",
    "        ## translate\n",
    "        current_end = aft.get_resi_vector(aft.start)['N']  # re-get as it moved.\n",
    "        missing = aft.end - fore.start\n",
    "        missing_d = abs(missing * 3.5) ** 0.5  #the James Ross fudge distance\n",
    "        if self.debug: print(f'Projecting {missing_d} + {extra_d} Å')\n",
    "        d = missing_d + internal_d + extra_d\n",
    "        target_end = np.array([d * np.cos(-θ)  * np.cos(φ) + ref.x,\n",
    "                               d * np.sin(-θ) * np.cos(φ) + ref.y,\n",
    "                               d * np.sin(-θ) * np.sin(φ) + ref.z])\n",
    "        pymol.cmd.translate((target_end - current_end).tolist(), aft.name, camera = 0)\n",
    "        if self.debug:\n",
    "            fore.add_NC_arrow('cylinderF', r=0, g=0, b=1)\n",
    "            aft.add_NC_arrow('cylinderA', r=0, g=1, b=0)\n",
    "            \n",
    "    def has_overlap(self, fore, aft):\n",
    "        o = pymol.cmd.overlap(f\"model {fore.name} and chain {fore.chain}\", f\"model {aft.name} and chain{aft.chain}\")\n",
    "        if o > 0:\n",
    "            if self.debug:\n",
    "                print(f\"Overlap between {fore.name} chain {fore.chain} and {aft.name} chain{aft.chain} of {o}\")\n",
    "            return True\n",
    "        o = pymol.cmd.overlap(f\"model {fore.name}\", f\"model {aft.name}\")\n",
    "        if o > 0:\n",
    "            if self.debug:\n",
    "                print(f\"Overlap between {fore.name} (whole) and {aft.name} (whole)\")\n",
    "            return True\n",
    "        return False\n",
    "            \n",
    "    \n",
    "    def add_xyz(self):\n",
    "        pymol.cmd.load_cgo( [ 9.0, 0,0,0, 100, 0, 0, 3, 1, 1, 1, 0, 0, 0], \"cylinderx\" )\n",
    "        pymol.cmd.load_cgo( [ 9.0, 0,0,0, 0, 100, 0, 3, 1, 1, 1, 0, 0, 0], \"cylindery\" )\n",
    "        pymol.cmd.load_cgo( [ 9.0, 0,0,0, 0, 0,100,  3, 1, 1, 1, 0, 0, 0], \"cylinderz\" )\n",
    "        \n",
    "    def safe_select(self, name, sele=None, silent=False):\n",
    "        \"\"\"\n",
    "        There is something not quite right with the selection. It throws an error...\"\"\"\n",
    "        try:\n",
    "            if not sele:\n",
    "                sele = name\n",
    "                name = 'test'\n",
    "            n = pymol.cmd.select(name, sele)\n",
    "            return n\n",
    "        except:\n",
    "            if '{' in sele:\n",
    "                raise ValueError(f'There seems to be a curly bracket in the selection ({sele}): bad formatting?')\n",
    "            else:\n",
    "                if not silent:\n",
    "                    warn(f'Failed selection \"{sele}\"')\n",
    "                return 0\n",
    "            \n",
    "###############################################################################################\n",
    "\n",
    "class Model:\n",
    "    def __init__(self, mode, chain, code, x, y, start=1, end=None, seq=None, length=None, workshop=None, tier=0, metadata=None, similars=None, subtender=None):\n",
    "        \"\"\"\n",
    "        Init just stores the variables. but .fetch_and_clean() does all the work by calling:\n",
    "        * model.kill_states()     ---> use only state 1.\n",
    "        * model.move_away_hetatm()  ---> protect the hetatms ligands!\n",
    "        * model.clean_Nterminus()   ---> trim the linkers!\n",
    "        * model.clean_Cterminus()\n",
    "        * model.fix_numbers()   --> the numbering off quite frequently...\n",
    "        * model.shift_to_true_Nterminus()  ---> gets rid of unsolved terminus\n",
    "        * model.shift_to_true_Cterminus()  ---> gets rid of unsolved terminus\n",
    "        While other methods are called by the project method of workshop. Say all the angle operations.\n",
    "        e.g.model.roll(n) or model.get_resi_coords(resi)\n",
    "        \n",
    "        The code can be a PDB code. or... PDB:code  or SWISSMODEL:id or LOCAL:filename (without underscores!)\n",
    "        \n",
    "        There are three sets of numbers.\n",
    "        uniprot start (`x`) which is the position in the whole protein of the first real residue based on the stated sequence regardless of density.\n",
    "        start (`start`) which is the first residue stated in PDBe query based on the stated sequence regardless of density.\n",
    "        Namely... the mapping start is at what residue index within the model does the match start,\n",
    "        while uniprot start is what the position that residues has in the whole sequence.\n",
    "        A method to keep an eye for is `.angle_fix()`. This ensures that the the N and C termini are on the x axis.\n",
    "        \"\"\"\n",
    "        self.chain = chain\n",
    "        self.mode = mode\n",
    "        self.code = code\n",
    "        self.name = code ## a backup value in case of duplicates...\n",
    "        self.tier = tier\n",
    "        self.metadata = metadata\n",
    "        self.x = x\n",
    "        self.y = y\n",
    "        self.start = start\n",
    "        if end:\n",
    "            self.end = end\n",
    "        else:\n",
    "            self.end = y - x\n",
    "        if length:\n",
    "            self.length = length\n",
    "        else:\n",
    "            self.length = y - x\n",
    "        self.seq = seq\n",
    "        self.subtender = subtender\n",
    "        if workshop:\n",
    "            self.workshop = workshop\n",
    "        else:\n",
    "            warn('No workshop?')\n",
    "            self.workshop = namedtuple('fakeshop', ['debug'])(False)\n",
    "        self.length = length\n",
    "        \n",
    "    @classmethod\n",
    "    def from_ds(cls, ds):\n",
    "        cls(**ds)\n",
    "        \n",
    "    def sele(self, resi, atom=None, no_hetatm=True):\n",
    "        if isinstance(resi, float):\n",
    "            resi = int(resi)\n",
    "        if no_hetatm:\n",
    "            extra = ' and not hetatm'\n",
    "        else:\n",
    "            extra = ''\n",
    "        if atom:\n",
    "            return f'/{self.name}//{self.chain}/{resi}/{atom}'+extra\n",
    "        else:\n",
    "            return f'/{self.name}//{self.chain}/{resi}/'+extra\n",
    "        \n",
    "    def __str__(self):\n",
    "        return f'<Model name={self.name} chain={self.chain} start={self.start} end={self.end}>'\n",
    "    \n",
    "    def describe(self):\n",
    "        (ref, pointer, θ, φ, internal_d) = self.angles\n",
    "        print(f'Angles: θ={np.rad2deg(θ)} and φ={np.rad2deg(φ)})')\n",
    "        print('Nterm:', self.get_resi_coords(self.start)['N'])\n",
    "        print('Cterm:', self.get_resi_coords(self.end)['C'])\n",
    "        \n",
    "    @staticmethod\n",
    "    def split_id(ID):\n",
    "        \"\"\"\n",
    "        splits an entry id into mode, code, chain\n",
    "        \"\"\"\n",
    "        if isinstance(ID, dict):\n",
    "            ID = ID['id']\n",
    "        if ':' not in ID:\n",
    "            mode = 'PDB'\n",
    "        else:\n",
    "            mode, ID = ID.split(':')\n",
    "        if '_' in ID:\n",
    "            code, chain = ID.split('_')\n",
    "        else:\n",
    "            code, chain = ID, 'A'\n",
    "        return mode, code, chain\n",
    "    \n",
    "    @classmethod\n",
    "    def infer_metadata(cls, ID):\n",
    "        mode, code, chain = cls.split(ID)\n",
    "        m = cls(mode=mode, chain=chain, code=code, start=1, end=1, x=1, y=1, seq='', length=1).load()\n",
    "        residex = m.resi_list()\n",
    "        pymol.delete(m.name)\n",
    "        return {'start': residex[0], 'end': residex[-1], 'length': residex[-1] - residex[0]}\n",
    "    \n",
    "    def resi_list(self):\n",
    "        return sorted(set([int(at.resi) for at in pymol.cmd.get_model(m.name).atom]))\n",
    "    \n",
    "    def residues(self):\n",
    "        rset = set([(at.resn, int(at.resi), True) for at in pymol.cmd.get_model(self.name).atom])\n",
    "        r = sorted(rset, key=lambda x: x[1])\n",
    "        s = ''.join(pymol.cmd.get_fastastr(f'{self.name} and chain {self.chain} and not hetatm').split('\\n')[1:])\n",
    "        first_stated_resn = seq3(s[0]).upper()\n",
    "        first_structured_resi = r[0][1]\n",
    "        first_structured_resn = r[0][0]\n",
    "        \n",
    "        if first_structured_resi == 1:\n",
    "            if first_stated_resn == first_structured_resn: ## case 1. all is good.\n",
    "                pass\n",
    "            else: ## case 2. negative numbers??!\n",
    "                raise NotImplementedError(f'First residues are {r[0:3]} but the sequence is {s}')\n",
    "        elif first_structured_resi > 1: \n",
    "            if first_stated_resn == first_structured_resn: ## case 3. The first residue exist is not 1.\n",
    "                pass\n",
    "            elif seq3(s[first_structured_resi - 1]).upper() == first_structured_resn:\n",
    "                ## case 4. the first stated residue is 1, but does not exist.\n",
    "                r = [(seq3(s[i]).upper(),i+1,False) for i in range(1, first_structured_resi)] + r\n",
    "            else:\n",
    "                raise NotImplementedError(f'First residues are {r[0:3]} but the sequence is {s}')\n",
    "        else:\n",
    "            raise NotImplementedError(f'First residues are {r[0:3]} but the sequence is {s}')\n",
    "    \n",
    "        return r        \n",
    "        \n",
    "    ###################### METHODS FOR LOADING ####################################\n",
    "        \n",
    "    def fetch_n_clean(self):\n",
    "        if self.workshop.debug: print(f'### Fetching for {self.name}')\n",
    "        c = self.code\n",
    "        self.load()\n",
    "        self.force_chain_A()\n",
    "        self.start_from_one()\n",
    "        self.move_away_hetatm()\n",
    "        self.clean_Nterminus()\n",
    "        self.clean_Cterminus()\n",
    "        self.fix_numbers()\n",
    "        self.shift_to_true_Nterminus()\n",
    "        self.shift_to_true_Cterminus()\n",
    "        \n",
    "    def load(self):\n",
    "        print(f'loading {self.name}')\n",
    "        if self.name in pymol.cmd.get_object_list():\n",
    "            preexisting = self.name\n",
    "            self.name+='X'\n",
    "            warn(f'{preexisting} is already loaded. Copying as {self.name}.')\n",
    "            pymol.cmd.create(self.name, preexisting)\n",
    "            return self\n",
    "        if self.mode == 'PDB':\n",
    "            print('Fetching biological assembly.')\n",
    "            self.name = pymol.cmd.fetch(self.code, type='pdb1')\n",
    "            if self.name == -1:\n",
    "                self.name = pymol.cmd.fetch(self.code)\n",
    "                print('\\t... solved by fetching assimetric unit.')\n",
    "            if self.name == -1:\n",
    "                raise ValueError(f'Could not load {self.code}')\n",
    "            self.kill_states()\n",
    "        elif self.mode == 'SWISSMODEL':\n",
    "            url = f'https://swissmodel.expasy.org/repository/{self.code}.pdb'\n",
    "            if self.workshop.debug:\n",
    "                print(url)\n",
    "            x = requests.get(url).text\n",
    "            if 'END' in x:\n",
    "                pymol.cmd.read_pdbstr( x, self.code)\n",
    "            else:\n",
    "                raise ValueError(f'{url} failed.')\n",
    "            chains = sorted({at.chain for at in pymol.cmd.get_model(self.name).atom})\n",
    "            if len(chains) == 1:\n",
    "                if chains[0] == '':\n",
    "                    pymol.alter.cmd(self.name, 'chain=\"A\"')\n",
    "                    self.chain = 'A'\n",
    "                else:\n",
    "                    self.chain = chains[0]\n",
    "            else:\n",
    "                raise NotImplementedError\n",
    "        elif self.mode == 'LOCAL':\n",
    "            pymol.cmd.load(self.code)\n",
    "            self.name = os.path.splitext(self.code)[0]\n",
    "        else:\n",
    "            raise ValueError(f'UNKNOWN mode {self.mode}. PDB: SWISSMODEL: LOCAL:')\n",
    "        if self.seq is None:\n",
    "            self.seq = ''.join(pymol.cmd.get_fastastr(f'{self.name} and chain {self.chain} and not hetatm').split('\\n')[1:])\n",
    "        return self\n",
    "    \n",
    "    def force_chain_A(self):\n",
    "        get_untaken = lambda c: list(set('ABCDEFGHIJKLMNOPQRSTUVWXYZ').difference(c))[0]\n",
    "        if self.chain != 'A':\n",
    "            chains = sorted({at.chain for at in pymol.cmd.get_model(self.name).atom})\n",
    "            if 'A' in chains:\n",
    "                # move chain A away.\n",
    "                new = get_untaken(chains)\n",
    "                pymol.cmd.alter(f'model {self.name} and chain A', f'chain=\"{new}\"')\n",
    "                pymol.cmd.sort()\n",
    "            pymol.cmd.alter(f'model {self.name} and chain {self.chain}', f'chain=\"A\"')\n",
    "            pymol.cmd.sort()\n",
    "        segis = sorted({at.chain for at in pymol.cmd.get_model(f'model {self.name} and chain A').atom})\n",
    "        if len(segis) > 1:\n",
    "            if '' not in segis:\n",
    "                subs = segis[1:]\n",
    "            else:\n",
    "                i = segis.index('')\n",
    "                subs = segis[:i]+segis[i+1:]\n",
    "            for s in subs:\n",
    "                new = get_untaken(chains)\n",
    "                pymol.cmd.alter(f'model {self.name} and chain A and segi {s}', f'chain=\"{new}\"')\n",
    "                pymol.cmd.sort()\n",
    "        elif segis[0] != '':\n",
    "            pymol.cmd.alter(f'model {self.name} and chain A and segi {segis[0]}', f'segi=\"\"')\n",
    "            pymol.cmd.sort()\n",
    "        else:\n",
    "            pass\n",
    "        self.chain = 'A'\n",
    "        return self\n",
    "                \n",
    "        \n",
    "    def kill_states(self):\n",
    "        #fetch(.., state=1) does not work. !?\n",
    "        pymol.cmd.split_states(self.name,1,1)\n",
    "        pymol.cmd.set_name(f'{self.name}_0001',self.code)\n",
    "        self.name = self.code\n",
    "        return self\n",
    "    \n",
    "    def move_away_hetatm(self):\n",
    "        if self.workshop.debug: print(f'### Protecting hetatms for {self.name}')\n",
    "        \"\"\"\n",
    "        Renumbers the hetatm residues to 9000+ values, unique in the whole workshop.\n",
    "        \"\"\"\n",
    "        hetatms = {self.sele(at.resi, no_hetatm=False) for at in pymol.cmd.get_model(self.name).atom if at.hetatm}\n",
    "        for h in hetatms:\n",
    "            pymol.cmd.alter(h, f\"resi={self.workshop.last_het_index}\")\n",
    "            self.workshop.last_het_index += 1\n",
    "        pymol.cmd.sort()\n",
    "        return self\n",
    "    \n",
    "        \n",
    "    def start_from_one(self):\n",
    "        \"\"\"\n",
    "        The first residue may exist or not.\n",
    "        \"\"\"\n",
    "        residues = self.residues()\n",
    "        if residues[0][1] != 1:\n",
    "            Δ = residues[0][1] - 1 \n",
    "            pymol.cmd.alter(f'{self.name} and chain {self.chain}',f'resi=str(int(resi)-{Δ})')\n",
    "            pymol.cmd.sort()\n",
    "        return self\n",
    "    \n",
    "    def clean_Nterminus(self):\n",
    "        if self.workshop.debug: print(f'### NTerm cleaning for {self.name}')\n",
    "        pymol.cmd.remove(self.sele(f'0-{self.start - 1}'))\n",
    "        pymol.cmd.sort()\n",
    "        return self\n",
    "    \n",
    "    def clean_Cterminus(self):\n",
    "        if self.workshop.debug: print(f'### CTerm cleaning for {self.name}')\n",
    "        \"\"\"\n",
    "        convert the N of the junk resi to OXT.\n",
    "        \"\"\"\n",
    "        has_trailing = pymol.cmd.alter( self.sele(self.end+1,'N') ,'name=\"OXT\"')\n",
    "        if has_trailing:\n",
    "            pymol.cmd.alter(self.sele(self.end+1,'N'), f'resi=\"{self.end}\"')\n",
    "            pymol.cmd.sort()\n",
    "            resn = pymol.cmd.get_model(self.sele(self.end)).atom[0].resn\n",
    "            pymol.cmd.alter(self.sele(self.end+1,'OXT'), f'resn=\"{resn}\"')\n",
    "            pymol.cmd.sort()\n",
    "        pymol.cmd.remove(self.sele(f'{self.end+1}-9999'))\n",
    "        pymol.cmd.sort()\n",
    "        return self\n",
    "        \n",
    "    def fix_numbers(self):\n",
    "        if self.workshop.debug:\n",
    "            print(f'### Fixing numbering for {self.name}')\n",
    "        if self.start != self.x:\n",
    "            Δ = self.x - self.start\n",
    "            pymol.cmd.alter(f'{self.name} and chain {self.chain}', f'resi=str(int(resi) + {Δ})')\n",
    "            pymol.cmd.sort()\n",
    "            self.start = self.start + Δ\n",
    "            self.end = self.end + Δ\n",
    "        return self\n",
    "            \n",
    "    def shift_to_true_Nterminus(self):\n",
    "        \"\"\"\n",
    "        The model may have missing residues at the NTerm, but in reality it is as if they were not there.\n",
    "        \"\"\"\n",
    "        if self.workshop.debug: print(f'### Shifting Nterm for {self.name}')\n",
    "        ## verify that the start exists.\n",
    "        coords = None\n",
    "        l = self.start\n",
    "        while coords is None:\n",
    "            coords = pymol.cmd.get_coords(self.sele(l))\n",
    "            l += 1\n",
    "        self.start = l -1\n",
    "\n",
    "    def shift_to_true_Cterminus(self):\n",
    "        \"\"\"\n",
    "        The model may have missing residues at the CTerm, but in reality it is as if they were not there.\n",
    "        \"\"\"\n",
    "        if self.workshop.debug: print(f'### Shifting Cterm for {self.name}')\n",
    "        ## verify that the end exists.\n",
    "        coords = None\n",
    "        l = self.end\n",
    "        while coords is None:\n",
    "            coords = pymol.cmd.get_coords(self.sele(l))\n",
    "            l -= 1\n",
    "            if l < 0:\n",
    "                raise ValueError(f'Trimming missing residues went haywire! start={self.x}/{self.start} end={self.y}/{self.end}. Died at {self.sele(l)}')\n",
    "        self.end = l + 1\n",
    "        \n",
    "    ###################### METHODS FOR ADJUSTING ####################################\n",
    "    def get_resi_coords(self, resi):\n",
    "        match_coords = lambda c: {'x': c[0],'y': c[1], 'z': c[2]}\n",
    "        coordsdex = self.get_resi_vector(resi)\n",
    "        return {atom: Coords(name= atom, **match_coords(coordsdex[atom]))\n",
    "                for atom in coordsdex}\n",
    "\n",
    "    def distance(self, A, B):\n",
    "        return ((A.x - B.x)**2 + (A.y - B.y)**2 + (A.z - B.z)**2 ) ** 0.5\n",
    "\n",
    "    def get_resi_vector(self, resi):\n",
    "        #if an error occurs here, it is because there is no self.sele(resi), which should not be possible\n",
    "        #as it is dealt with upstream in the trimming. Export the scene with pymol.exporting.multisave('tmp.pdb')\n",
    "        #as see what is wrong. It is somethign weird. Maybe a hetatm not marked as such?\n",
    "        return {atom: pymol.cmd.get_coords(self.sele(resi, atom))[0] for atom in ('C', 'N', 'CA')}\n",
    "\n",
    "    @property\n",
    "    def angles(self):\n",
    "        ref = self.get_resi_coords(self.start)['N']\n",
    "        pointer = self.get_resi_coords(self.end)['C']\n",
    "        internal_d = self.distance(pointer, ref)\n",
    "        θ = np.arctan((pointer.y - ref.y)/(pointer.x - ref.x))\n",
    "        φ = np.arctan((pointer.z - ref.z)/(pointer.x - ref.x))\n",
    "        return (ref, pointer, θ, φ, internal_d)\n",
    "\n",
    "    def add_NC_arrow(self, name, r=0, g=0, b=0):\n",
    "        N = self.get_resi_coords(self.start)['N']\n",
    "        C = self.get_resi_coords(self.end)['C']\n",
    "        pymol.cmd.load_cgo( [ 9.0, N.x, N.y, N.z, C.x, C.y, C.z, 3, r, g, b, r/3, g/3, b/3], name )\n",
    "\n",
    "    def camera_fix(self):\n",
    "        #pymol.cmd.alter('all','segi=\"\"')\n",
    "        self.angle_fix()\n",
    "\n",
    "    def angle_fix(self, cycles=3):\n",
    "        if self.workshop.debug:\n",
    "            print(f'Initial position for model {self.name}.....')\n",
    "            self.describe()\n",
    "        ## rotate the xy plane\n",
    "        for i in range(cycles):\n",
    "            (ref, pointer, θ, φ, internal_d) = self.angles\n",
    "            pymol.cmd.translate([-ref.x, -ref.y, -ref.z], self.name, camera = 0)\n",
    "            pymol.cmd.rotate([0, 0, 1], -np.rad2deg(θ), selection=self.name, camera = 0)\n",
    "            ## rotate the xz plane\n",
    "            (ref, pointer, θ, φ, internal_d) = self.angles\n",
    "            pymol.cmd.translate([-ref.x, -ref.y, -ref.z], self.name, camera = 0)\n",
    "            pymol.cmd.rotate([0, 1, 0], np.rad2deg(φ), selection=self.name, camera = 0)\n",
    "            (ref, pointer, θ, φ, internal_d) = self.angles\n",
    "            pymol.cmd.translate([-ref.x, -ref.y, -ref.z], self.name, camera = 0)\n",
    "            tail = self.get_resi_coords(self.end)['C']\n",
    "            if tail.x < 0:\n",
    "                if self.workshop.debug:\n",
    "                    print('It is off by 180... Odd.')\n",
    "                pymol.cmd.rotate([0, 0, 1], 180, selection=self.name, camera = 0) ## rotate the xy plane\n",
    "                (ref, pointer, θ, φ, internal_d) = self.angles\n",
    "                pymol.cmd.translate([-ref.x, -ref.y, -ref.z], self.name, camera = 0)\n",
    "        if self.workshop.debug:\n",
    "            print(f'Final position for model {self.name}.....')\n",
    "            self.describe()\n",
    "        return self\n",
    "            \n",
    "    def roll(self, φ):\n",
    "        \"\"\"\n",
    "        Whereas they are aligned on the x-axis with the y,z coordinates of C and N termini near 0,0.\n",
    "        They can rotate around the X. roll. This is good to remove clashes.\n",
    "        \"\"\"\n",
    "        pymol.cmd.rotate([1, 0, 0], φ, selection=self.name, camera = 0)\n",
    "        self.angle_fix()\n",
    "        return self"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "## vigilin\n",
    "uniprot = 'Q00341'\n",
    "pdbs = [{'x': 432, 'y': 502, 'id': '1VIG_A', 'description': '1VIG'}, \n",
    "        {'x': 432, 'y': 502, 'id': '1VIH_A', 'description': '1VIH'}, \n",
    "        {'x': 142, 'y': 222, 'id': '2CTE_A', 'description': '2CTE'},\n",
    "        {'x': 346, 'y': 434, 'id': '2CTF_A', 'description': '2CTF'}, \n",
    "        {'x': 645, 'y': 726, 'id': '2CTJ_A', 'description': '2CTJ'},\n",
    "        {'x': 964, 'y': 1054, 'id': '2CTK_A', 'description': '2CTK'},\n",
    "        {'x': 1044, 'y': 1127, 'id': '2CTL_A', 'description': '2CTL'},\n",
    "        {'x': 1119, 'y': 1200, 'id': '2CTM_A', 'description': '2CTM'}]\n",
    "swissmodel = [{'x': 581, 'y': 724, 'id': '5d5620399ffd12cf2f734256', 'description': '2n8m.1.A (id:2e+01%)'},\n",
    "              {'x': 1053, 'y': 1198, 'id': '5c5f33ed8fd6f9f9ae62f475', 'description': '2jvz.1.A (id:2e+01%)'},\n",
    "              {'x': 583, 'y': 723, 'id': '5c5f33ed8fd6f9f9ae62f465', 'description': '3aev.1.B (id:3e+01%)'},\n",
    "              {'x': 726, 'y': 1068, 'id': '5d5620399ffd12cf2f73425e', 'description': '3n89.1.A (id:1e+01%)'},\n",
    "              {'x': 346, 'y': 434, 'id': '5c5f33ed8fd6f9f9ae62f469', 'description': '2ctf.1.A (id:1e+02%)'},\n",
    "              {'x': 1053, 'y': 1199, 'id': '5d56203a9ffd12cf2f734266', 'description': '3krm.1.A (id:2e+01%)'},\n",
    "              {'x': 508, 'y': 648, 'id': '5d5620399ffd12cf2f734246', 'description': '5wyj.31.A (id:2e+01%)'},\n",
    "              {'x': 584, 'y': 720, 'id': '5c5f33ed8fd6f9f9ae62f485', 'description': '1j4w.1.B (id:2e+01%)'},\n",
    "              {'x': 860, 'y': 1194, 'id': '5d5620399ffd12cf2f73424e', 'description': '3n89.1.A (id:2e+01%)'},\n",
    "              {'x': 655, 'y': 798, 'id': '5c5f33ed8fd6f9f9ae62f491', 'description': '2jvz.1.A (id:2e+01%)'},\n",
    "              {'x': 653, 'y': 796, 'id': '5c5f33ed8fd6f9f9ae62f499', 'description': '6fai.1.F (id:2e+01%)'},\n",
    "              {'x': 651, 'y': 799, 'id': '5d5620399ffd12cf2f73425a', 'description': '2n8m.1.A (id:2e+01%)'},\n",
    "              {'x': 73, 'y': 219, 'id': '5d5620399ffd12cf2f73423a', 'description': '3i82.1.A (id:1e+01%)'},\n",
    "              {'x': 1055, 'y': 1193, 'id': '5c5f33ed8fd6f9f9ae62f471', 'description': '1j4w.1.B (id:2e+01%)'},\n",
    "              {'x': 432, 'y': 502, 'id': '5be4a0a502efd0cd3370539f', 'description': '1vih (id:%)'},\n",
    "              {'x': 432, 'y': 502, 'id': '5be49e6e02efd0c70a281b5c', 'description': '1vig (id:%)'},\n",
    "              {'x': 346, 'y': 434, 'id': '5be49cbe02efd0c17498a975', 'description': '2ctf (id:%)'},\n",
    "              {'x': 1119, 'y': 1200, 'id': '5be4adbd02efd0f036d2a8c4', 'description': '2ctm (id:%)'},\n",
    "               {'x': 964, 'y': 1054, 'id': '5be4a89802efd0e17f75807a', 'description': '2ctk (id:%)'},\n",
    "              {'x': 1044, 'y': 1127, 'id': '5be4ab5f02efd0e942407c0c', 'description': '2ctl (id:%)'},\n",
    "              {'x': 142, 'y': 222, 'id': '5be4999502efd0b97cdb5dd6', 'description': '2cte (id:%)'},\n",
    "              {'x': 645, 'y': 726, 'id': '5be4a62b02efd0db3ee53fd0', 'description': '2ctj (id:%)'}]\n",
    "features = {'initiator methionine': [{'x': 1, 'y': 1, 'description': 'Removed', 'id': 'initiatormethionine_1', 'type': 'initiator methionine'}], 'chain': [{'x': 2, 'y': 1268, 'description': 'Vigilin', 'id': 'chain_2_1268', 'type': 'chain'}], 'domain': [{'x': 158, 'y': 229, 'description': 'KH 1', 'id': 'domain_158_229', 'type': 'domain'}, {'x': 230, 'y': 302, 'description': 'KH 2', 'id': 'domain_230_302', 'type': 'domain'}, {'x': 303, 'y': 371, 'description': 'KH 3', 'id': 'domain_303_371', 'type': 'domain'}, {'x': 372, 'y': 442, 'description': 'KH 4', 'id': 'domain_372_442', 'type': 'domain'}, {'x': 443, 'y': 514, 'description': 'KH 5', 'id': 'domain_443_514', 'type': 'domain'}, {'x': 515, 'y': 588, 'description': 'KH 6', 'id': 'domain_515_588', 'type': 'domain'}, {'x': 589, 'y': 660, 'description': 'KH 7', 'id': 'domain_589_660', 'type': 'domain'}, {'x': 661, 'y': 734, 'description': 'KH 8', 'id': 'domain_661_734', 'type': 'domain'}, {'x': 735, 'y': 807, 'description': 'KH 9', 'id': 'domain_735_807', 'type': 'domain'}, {'x': 808, 'y': 880, 'description': 'KH 10', 'id': 'domain_808_880', 'type': 'domain'}, {'x': 881, 'y': 979, 'description': 'KH 11', 'id': 'domain_881_979', 'type': 'domain'}, {'x': 980, 'y': 1059, 'description': 'KH 12', 'id': 'domain_980_1059', 'type': 'domain'}, {'x': 1060, 'y': 1134, 'description': 'KH 13', 'id': 'domain_1060_1134', 'type': 'domain'}, {'x': 1135, 'y': 1209, 'description': 'KH 14', 'id': 'domain_1135_1209', 'type': 'domain'}], 'modified residue': [{'x': 2, 'y': 2, 'description': 'N-acetylserine', 'id': 'modifiedresidue_2', 'type': 'modified residue'}, {'x': 8, 'y': 8, 'description': 'Phosphothreonine', 'id': 'modifiedresidue_8', 'type': 'modified residue'}, {'x': 11, 'y': 11, 'description': 'Phosphoserine', 'id': 'modifiedresidue_11', 'type': 'modified residue'}, {'x': 31, 'y': 31, 'description': 'Phosphoserine', 'id': 'modifiedresidue_31', 'type': 'modified residue'}, {'x': 35, 'y': 35, 'description': 'Phosphoserine', 'id': 'modifiedresidue_35', 'type': 'modified residue'}, {'x': 295, 'y': 295, 'description': 'Phosphothreonine', 'id': 'modifiedresidue_295', 'type': 'modified residue'}, {'x': 296, 'y': 296, 'description': 'Phosphothreonine', 'id': 'modifiedresidue_296', 'type': 'modified residue'}, {'x': 317, 'y': 317, 'description': 'Phosphoserine', 'id': 'modifiedresidue_317', 'type': 'modified residue'}, {'x': 437, 'y': 437, 'description': 'Phosphotyrosine', 'id': 'modifiedresidue_437', 'type': 'modified residue'}, {'x': 645, 'y': 645, 'description': 'Phosphoserine', 'id': 'modifiedresidue_645', 'type': 'modified residue'}, {'x': 991, 'y': 991, 'description': 'N6-acetyllysine', 'id': 'modifiedresidue_991', 'type': 'modified residue'}, {'x': 1247, 'y': 1247, 'description': 'Phosphoserine', 'id': 'modifiedresidue_1247', 'type': 'modified residue'}, {'x': 1252, 'y': 1252, 'description': 'Phosphoserine', 'id': 'modifiedresidue_1252', 'type': 'modified residue'}], 'splice variant': [{'x': 1, 'y': 1, 'description': 'In isoform 2.', 'id': 'splicevariant_1', 'type': 'splice variant'}, {'x': 291, 'y': 359, 'description': 'In isoform 2.', 'id': 'splicevariant_291_359', 'type': 'splice variant'}], 'sequence variant': [{'x': 61, 'y': 61, 'description': 'In dbSNP:rs11891776.', 'id': 'sequencevariant_61', 'type': 'sequence variant'}, {'x': 229, 'y': 229, 'description': 'In dbSNP:rs7572799.', 'id': 'sequencevariant_229', 'type': 'sequence variant'}, {'x': 418, 'y': 418, 'description': 'In dbSNP:rs7578199.', 'id': 'sequencevariant_418', 'type': 'sequence variant'}, {'x': 568, 'y': 568, 'description': 'In a breast cancer sample; somatic mutation.', 'id': 'sequencevariant_568', 'type': 'sequence variant'}, {'x': 939, 'y': 939, 'description': 'In a breast cancer sample; somatic mutation.', 'id': 'sequencevariant_939', 'type': 'sequence variant'}, {'x': 1264, 'y': 1264, 'description': 'In dbSNP:rs12281.', 'id': 'sequencevariant_1264', 'type': 'sequence variant'}], 'strand': [{'x': 151, 'y': 156, 'description': 'strand', 'id': 'strand_151_156', 'type': 'strand'}, {'x': 168, 'y': 170, 'description': 'strand', 'id': 'strand_168_170', 'type': 'strand'}, {'x': 195, 'y': 200, 'description': 'strand', 'id': 'strand_195_200', 'type': 'strand'}, {'x': 365, 'y': 370, 'description': 'strand', 'id': 'strand_365_370', 'type': 'strand'}, {'x': 395, 'y': 401, 'description': 'strand', 'id': 'strand_395_401', 'type': 'strand'}, {'x': 403, 'y': 405, 'description': 'strand', 'id': 'strand_403_405', 'type': 'strand'}, {'x': 407, 'y': 412, 'description': 'strand', 'id': 'strand_407_412', 'type': 'strand'}, {'x': 436, 'y': 442, 'description': 'strand', 'id': 'strand_436_442', 'type': 'strand'}, {'x': 468, 'y': 470, 'description': 'strand', 'id': 'strand_468_470', 'type': 'strand'}, {'x': 476, 'y': 487, 'description': 'strand', 'id': 'strand_476_487', 'type': 'strand'}, {'x': 656, 'y': 659, 'description': 'strand', 'id': 'strand_656_659', 'type': 'strand'}, {'x': 671, 'y': 673, 'description': 'strand', 'id': 'strand_671_673', 'type': 'strand'}, {'x': 687, 'y': 689, 'description': 'strand', 'id': 'strand_687_689', 'type': 'strand'}, {'x': 699, 'y': 704, 'description': 'strand', 'id': 'strand_699_704', 'type': 'strand'}, {'x': 973, 'y': 978, 'description': 'strand', 'id': 'strand_973_978', 'type': 'strand'}, {'x': 990, 'y': 992, 'description': 'strand', 'id': 'strand_990_992', 'type': 'strand'}, {'x': 1005, 'y': 1007, 'description': 'strand', 'id': 'strand_1005_1007', 'type': 'strand'}, {'x': 1017, 'y': 1022, 'description': 'strand', 'id': 'strand_1017_1022', 'type': 'strand'}, {'x': 1054, 'y': 1058, 'description': 'strand', 'id': 'strand_1054_1058', 'type': 'strand'}, {'x': 1070, 'y': 1072, 'description': 'strand', 'id': 'strand_1070_1072', 'type': 'strand'}, {'x': 1085, 'y': 1087, 'description': 'strand', 'id': 'strand_1085_1087', 'type': 'strand'}, {'x': 1098, 'y': 1105, 'description': 'strand', 'id': 'strand_1098_1105', 'type': 'strand'}, {'x': 1130, 'y': 1133, 'description': 'strand', 'id': 'strand_1130_1133', 'type': 'strand'}, {'x': 1145, 'y': 1147, 'description': 'strand', 'id': 'strand_1145_1147', 'type': 'strand'}, {'x': 1160, 'y': 1162, 'description': 'strand', 'id': 'strand_1160_1162', 'type': 'strand'}, {'x': 1173, 'y': 1178, 'description': 'strand', 'id': 'strand_1173_1178', 'type': 'strand'}], 'turn': [{'x': 159, 'y': 161, 'description': 'turn', 'id': 'turn_159_161', 'type': 'turn'}, {'x': 381, 'y': 384, 'description': 'turn', 'id': 'turn_381_384', 'type': 'turn'}, {'x': 452, 'y': 454, 'description': 'turn', 'id': 'turn_452_454', 'type': 'turn'}, {'x': 693, 'y': 696, 'description': 'turn', 'id': 'turn_693_696', 'type': 'turn'}, {'x': 1011, 'y': 1013, 'description': 'turn', 'id': 'turn_1011_1013', 'type': 'turn'}, {'x': 1061, 'y': 1063, 'description': 'turn', 'id': 'turn_1061_1063', 'type': 'turn'}, {'x': 1091, 'y': 1093, 'description': 'turn', 'id': 'turn_1091_1093', 'type': 'turn'}, {'x': 1136, 'y': 1138, 'description': 'turn', 'id': 'turn_1136_1138', 'type': 'turn'}], 'helix': [{'x': 162, 'y': 166, 'description': 'helix', 'id': 'helix_162_166', 'type': 'helix'}, {'x': 172, 'y': 179, 'description': 'helix', 'id': 'helix_172_179', 'type': 'helix'}, {'x': 202, 'y': 220, 'description': 'helix', 'id': 'helix_202_220', 'type': 'helix'}, {'x': 376, 'y': 380, 'description': 'helix', 'id': 'helix_376_380', 'type': 'helix'}, {'x': 386, 'y': 393, 'description': 'helix', 'id': 'helix_386_393', 'type': 'helix'}, {'x': 414, 'y': 434, 'description': 'helix', 'id': 'helix_414_434', 'type': 'helix'}, {'x': 446, 'y': 450, 'description': 'helix', 'id': 'helix_446_450', 'type': 'helix'}, {'x': 457, 'y': 464, 'description': 'helix', 'id': 'helix_457_464', 'type': 'helix'}, {'x': 488, 'y': 499, 'description': 'helix', 'id': 'helix_488_499', 'type': 'helix'}, {'x': 662, 'y': 669, 'description': 'helix', 'id': 'helix_662_669', 'type': 'helix'}, {'x': 674, 'y': 683, 'description': 'helix', 'id': 'helix_674_683', 'type': 'helix'}, {'x': 706, 'y': 723, 'description': 'helix', 'id': 'helix_706_723', 'type': 'helix'}, {'x': 981, 'y': 988, 'description': 'helix', 'id': 'helix_981_988', 'type': 'helix'}, {'x': 993, 'y': 1001, 'description': 'helix', 'id': 'helix_993_1001', 'type': 'helix'}, {'x': 1024, 'y': 1050, 'description': 'helix', 'id': 'helix_1024_1050', 'type': 'helix'}, {'x': 1064, 'y': 1067, 'description': 'helix', 'id': 'helix_1064_1067', 'type': 'helix'}, {'x': 1074, 'y': 1082, 'description': 'helix', 'id': 'helix_1074_1082', 'type': 'helix'}, {'x': 1107, 'y': 1126, 'description': 'helix', 'id': 'helix_1107_1126', 'type': 'helix'}, {'x': 1139, 'y': 1143, 'description': 'helix', 'id': 'helix_1139_1143', 'type': 'helix'}, {'x': 1149, 'y': 1157, 'description': 'helix', 'id': 'helix_1149_1157', 'type': 'helix'}, {'x': 1180, 'y': 1198, 'description': 'helix', 'id': 'helix_1180_1198', 'type': 'helix'}]}\n",
    "\n",
    "p = [{'tier': 1, **p} for p in pdbs] + [{'tier': 2, **p, 'id': 'SWISSMODEL:'+p['id']} for p in swissmodel]   \n",
    "w = Workshop(p, debug=False).order()\n",
    "\n",
    "print(w.joints)\n",
    "\n",
    "pymol.exporting.multisave('tmp.pdb')\n",
    "\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 251,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>aft_resi_start</th>\n",
       "      <th>aft_source</th>\n",
       "      <th>distance</th>\n",
       "      <th>fore_resi_end</th>\n",
       "      <th>fore_resi_start</th>\n",
       "      <th>fore_source</th>\n",
       "      <th>overlap</th>\n",
       "      <th>type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>73.0</td>\n",
       "      <td>PDB:2CTE</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>end</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>150.0</td>\n",
       "      <td>PDB:2CTE</td>\n",
       "      <td>1.327132</td>\n",
       "      <td>149.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SWISSMODEL:5d5620399ffd12cf2f73423a</td>\n",
       "      <td>5.496016</td>\n",
       "      <td>alignment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>346.0</td>\n",
       "      <td>PDB:1VIG</td>\n",
       "      <td>38.749191</td>\n",
       "      <td>222.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDB:2CTE</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>projection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>434.0</td>\n",
       "      <td>PDB:1VIG</td>\n",
       "      <td>1.325057</td>\n",
       "      <td>433.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDB:2CTF</td>\n",
       "      <td>7.572303</td>\n",
       "      <td>alignment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>508.0</td>\n",
       "      <td>PDB:2CTM</td>\n",
       "      <td>54.671745</td>\n",
       "      <td>502.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDB:1VIG</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>projection</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>600.0</td>\n",
       "      <td>SWISSMODEL:5d5620399ffd12cf2f734256</td>\n",
       "      <td>1.353279</td>\n",
       "      <td>599.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SWISSMODEL:5d5620399ffd12cf2f734246</td>\n",
       "      <td>2.721029</td>\n",
       "      <td>alignment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>649.0</td>\n",
       "      <td>PDB:2CTJ</td>\n",
       "      <td>1.325627</td>\n",
       "      <td>648.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SWISSMODEL:5d5620399ffd12cf2f734256</td>\n",
       "      <td>4.157502</td>\n",
       "      <td>alignment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>727.0</td>\n",
       "      <td>SWISSMODEL:5d5620399ffd12cf2f73425a</td>\n",
       "      <td>1.366165</td>\n",
       "      <td>726.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDB:2CTJ</td>\n",
       "      <td>2.550897</td>\n",
       "      <td>alignment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>747.0</td>\n",
       "      <td>SWISSMODEL:5d5620399ffd12cf2f73425e</td>\n",
       "      <td>1.361739</td>\n",
       "      <td>746.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SWISSMODEL:5d5620399ffd12cf2f73425a</td>\n",
       "      <td>5.387156</td>\n",
       "      <td>alignment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>778.0</td>\n",
       "      <td>SWISSMODEL:5d5620399ffd12cf2f73425e</td>\n",
       "      <td>1.353988</td>\n",
       "      <td>777.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDB:2CTK</td>\n",
       "      <td>2.628903</td>\n",
       "      <td>alignment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1045.0</td>\n",
       "      <td>SWISSMODEL:5d5620399ffd12cf2f73424e</td>\n",
       "      <td>1.335035</td>\n",
       "      <td>1044.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SWISSMODEL:5d5620399ffd12cf2f73425e</td>\n",
       "      <td>2.768052</td>\n",
       "      <td>alignment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1050.0</td>\n",
       "      <td>SWISSMODEL:5d5620399ffd12cf2f73424e</td>\n",
       "      <td>1.342978</td>\n",
       "      <td>1049.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDB:2CTL</td>\n",
       "      <td>2.499620</td>\n",
       "      <td>alignment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1054.0</td>\n",
       "      <td>PDB:2CTK</td>\n",
       "      <td>1.366031</td>\n",
       "      <td>1053.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SWISSMODEL:5d5620399ffd12cf2f73425e</td>\n",
       "      <td>7.734130</td>\n",
       "      <td>alignment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1094.0</td>\n",
       "      <td>SWISSMODEL:5d56203a9ffd12cf2f734266</td>\n",
       "      <td>1.358572</td>\n",
       "      <td>1093.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SWISSMODEL:5d5620399ffd12cf2f73424e</td>\n",
       "      <td>2.776227</td>\n",
       "      <td>alignment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1127.0</td>\n",
       "      <td>PDB:2CTL</td>\n",
       "      <td>1.349044</td>\n",
       "      <td>1126.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SWISSMODEL:5d5620399ffd12cf2f73424e</td>\n",
       "      <td>4.142237</td>\n",
       "      <td>alignment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1197.0</td>\n",
       "      <td>PDB:2CTM</td>\n",
       "      <td>1.352944</td>\n",
       "      <td>1196.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>SWISSMODEL:5d56203a9ffd12cf2f734266</td>\n",
       "      <td>24.895792</td>\n",
       "      <td>alignment</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1200.0</td>\n",
       "      <td>PDB:2CTM</td>\n",
       "      <td>NaN</td>\n",
       "      <td>end</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    aft_resi_start                           aft_source   distance  \\\n",
       "15            73.0                             PDB:2CTE        NaN   \n",
       "0            150.0                             PDB:2CTE   1.327132   \n",
       "13           346.0                             PDB:1VIG  38.749191   \n",
       "1            434.0                             PDB:1VIG   1.325057   \n",
       "14           508.0                             PDB:2CTM  54.671745   \n",
       "2            600.0  SWISSMODEL:5d5620399ffd12cf2f734256   1.353279   \n",
       "3            649.0                             PDB:2CTJ   1.325627   \n",
       "4            727.0  SWISSMODEL:5d5620399ffd12cf2f73425a   1.366165   \n",
       "5            747.0  SWISSMODEL:5d5620399ffd12cf2f73425e   1.361739   \n",
       "7            778.0  SWISSMODEL:5d5620399ffd12cf2f73425e   1.353988   \n",
       "8           1045.0  SWISSMODEL:5d5620399ffd12cf2f73424e   1.335035   \n",
       "10          1050.0  SWISSMODEL:5d5620399ffd12cf2f73424e   1.342978   \n",
       "6           1054.0                             PDB:2CTK   1.366031   \n",
       "11          1094.0  SWISSMODEL:5d56203a9ffd12cf2f734266   1.358572   \n",
       "9           1127.0                             PDB:2CTL   1.349044   \n",
       "12          1197.0                             PDB:2CTM   1.352944   \n",
       "16             NaN                                  NaN        NaN   \n",
       "\n",
       "    fore_resi_end  fore_resi_start                          fore_source  \\\n",
       "15            NaN              NaN                                  NaN   \n",
       "0           149.0              NaN  SWISSMODEL:5d5620399ffd12cf2f73423a   \n",
       "13          222.0              NaN                             PDB:2CTE   \n",
       "1           433.0              NaN                             PDB:2CTF   \n",
       "14          502.0              NaN                             PDB:1VIG   \n",
       "2           599.0              NaN  SWISSMODEL:5d5620399ffd12cf2f734246   \n",
       "3           648.0              NaN  SWISSMODEL:5d5620399ffd12cf2f734256   \n",
       "4           726.0              NaN                             PDB:2CTJ   \n",
       "5           746.0              NaN  SWISSMODEL:5d5620399ffd12cf2f73425a   \n",
       "7           777.0              NaN                             PDB:2CTK   \n",
       "8          1044.0              NaN  SWISSMODEL:5d5620399ffd12cf2f73425e   \n",
       "10         1049.0              NaN                             PDB:2CTL   \n",
       "6          1053.0              NaN  SWISSMODEL:5d5620399ffd12cf2f73425e   \n",
       "11         1093.0              NaN  SWISSMODEL:5d5620399ffd12cf2f73424e   \n",
       "9          1126.0              NaN  SWISSMODEL:5d5620399ffd12cf2f73424e   \n",
       "12         1196.0              NaN  SWISSMODEL:5d56203a9ffd12cf2f734266   \n",
       "16            NaN           1200.0                             PDB:2CTM   \n",
       "\n",
       "      overlap        type  \n",
       "15        NaN         end  \n",
       "0    5.496016   alignment  \n",
       "13   0.000000  projection  \n",
       "1    7.572303   alignment  \n",
       "14   0.000000  projection  \n",
       "2    2.721029   alignment  \n",
       "3    4.157502   alignment  \n",
       "4    2.550897   alignment  \n",
       "5    5.387156   alignment  \n",
       "7    2.628903   alignment  \n",
       "8    2.768052   alignment  \n",
       "10   2.499620   alignment  \n",
       "6    7.734130   alignment  \n",
       "11   2.776227   alignment  \n",
       "9    4.142237   alignment  \n",
       "12  24.895792   alignment  \n",
       "16        NaN         end  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(pd.DataFrame.from_records(w.joints).sort_values(by='aft_resi_start'))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "* biological assembly...\n",
    "* verify that shift is okay."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Partner explorer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Cell division control protein 42 homolog</th>\n",
       "      <th>Dedicator of cytokinesis protein 9</th>\n",
       "      <th>N_other</th>\n",
       "      <th>N_peptides</th>\n",
       "      <th>code</th>\n",
       "      <th>nonprotein_details</th>\n",
       "      <th>protein_details</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1WG7</td>\n",
       "      <td>[]</td>\n",
       "      <td>[Dedicator of cytokinesis protein 9 (['A']/['A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2WM9</td>\n",
       "      <td>[GLYCEROL (['A', 'B']/['C', 'D', 'E']), water ...</td>\n",
       "      <td>[Dedicator of cytokinesis protein 9 (['A']/['A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2WMN</td>\n",
       "      <td>[GUANOSINE-5'-DIPHOSPHATE (['B']/['C']), water...</td>\n",
       "      <td>[Dedicator of cytokinesis protein 9 (['A']/['A...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>2WMO</td>\n",
       "      <td>[GUANOSINE-5'-TRIPHOSPHATE (['B']/['C']), MAGN...</td>\n",
       "      <td>[Dedicator of cytokinesis protein 9 (['A']/['A...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Cell division control protein 42 homolog  \\\n",
       "0                                       NaN   \n",
       "1                                       1.0   \n",
       "2                                       1.0   \n",
       "3                                       1.0   \n",
       "\n",
       "   Dedicator of cytokinesis protein 9  N_other  N_peptides  code  \\\n",
       "0                                   1        0           1  1WG7   \n",
       "1                                   1        2           2  2WM9   \n",
       "2                                   1        2           2  2WMN   \n",
       "3                                   1        3           2  2WMO   \n",
       "\n",
       "                                  nonprotein_details  \\\n",
       "0                                                 []   \n",
       "1  [GLYCEROL (['A', 'B']/['C', 'D', 'E']), water ...   \n",
       "2  [GUANOSINE-5'-DIPHOSPHATE (['B']/['C']), water...   \n",
       "3  [GUANOSINE-5'-TRIPHOSPHATE (['B']/['C']), MAGN...   \n",
       "\n",
       "                                     protein_details  \n",
       "0  [Dedicator of cytokinesis protein 9 (['A']/['A...  \n",
       "1  [Dedicator of cytokinesis protein 9 (['A']/['A...  \n",
       "2  [Dedicator of cytokinesis protein 9 (['A']/['A...  \n",
       "3  [Dedicator of cytokinesis protein 9 (['A']/['A...  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## list the partners.\n",
    "tdata = []\n",
    "for code, entry in data.items():\n",
    "    #print(entity['molecule_name'], entity['in_chains'], entity['in_struct_asyms'])\n",
    "    proteins = [entity for entity in entry if entity['molecule_type'] == 'polypeptide(L)']\n",
    "    protein_details = [f'{p[\"molecule_name\"][0]} ({p[\"in_chains\"]}/{p[\"in_struct_asyms\"]})' for p in proteins]\n",
    "                   \n",
    "    nonproteins = [entity for entity in entry if entity['molecule_type'] != 'polypeptide(L)']\n",
    "    nonprotein_details = [f'{p[\"molecule_name\"][0]} ({p[\"in_chains\"]}/{p[\"in_struct_asyms\"]})' for p in nonproteins]\n",
    "    #print(f'{code} has {len(proteins)} polypeptide chains: {protein_details}')\n",
    "    #print(f'{code} has {len(nonproteins)} non-polypeptide chains: {nonprotein_details}')\n",
    "    tdata.append({'code': code,\n",
    "         'N_peptides': len(proteins),\n",
    "         'N_other': len(nonproteins),\n",
    "         'protein_details': protein_details,\n",
    "         'nonprotein_details': nonprotein_details,\n",
    "         **{p[\"molecule_name\"][0]: len(p[\"in_chains\"]) for p in proteins}\n",
    "         })\n",
    "                          \n",
    "pd.DataFrame.from_records(tdata)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
